src/utils/time.ts

```diff
@@ -12,45 +22,93 @@ export interface Time {
+  const [hourString, minuteString, secondString] = value.split(":");
+  const hour = formatTimePart(parseInt(hourString));
+  const minute = formatTimePart(parseInt(minuteString));
+  if (secondString) {
+    const second = formatTimePart(parseInt(secondString));
+    return `${hour}:${minute}:${second}`;
   }
-  return hourAsNumber > 12 ? formatTimePart(hourAsNumber - 12) : hour;
+  return `${hour}:${minute}`;
 }
 
-export function parseTimeString(value: string): Time {
-  const timeString = formatTimeString(value);
-  const [hour, minute, second] = timeString ? timeString.split(":") : [null, null, null];
-  return {
-    hour,
-    minute,
-    second: second || (hour && minute ? "00" : null)
-  };
+export function getLocaleHourCycle(locale: string): HourCycle {
+  const formatter = createLocaleDateTimeFormatter(locale);
+  const parts = formatter.formatToParts(new Date(Date.UTC(0, 0, 0, 0, 0, 0)));
+  return getLocalizedTimePart("meridiem", parts) ? "12" : "24";
 }
 
-export function formatTimeString(value: string): string {
-  if (!value || value.endsWith(":") || value.startsWith(":")) {
+function getLocalizedTimePart(part: TimePart, parts: Intl.DateTimeFormatPart[]): string {
+  if (!part || !parts) {
     return null;
   }
+  if (part === "hourSuffix") {
+    const hourIndex = parts.indexOf(parts.find(({ type }): boolean => type === "hour"));
+    const minuteIndex = parts.indexOf(parts.find(({ type }): boolean => type === "minute"));
+    const hourSuffix = parts[hourIndex + 1];
+    return hourSuffix && hourSuffix.type === "literal" && minuteIndex - hourIndex === 2
+      ? hourSuffix.value?.trim() || null
+      : null;
+  }
+  if (part === "minuteSuffix") {
+    const minuteIndex = parts.indexOf(parts.find(({ type }): boolean => type === "minute"));
+    const secondIndex = parts.indexOf(parts.find(({ type }): boolean => type === "second"));
+    const minuteSuffix = parts[minuteIndex + 1];
+    return minuteSuffix && minuteSuffix.type === "literal" && secondIndex - minuteIndex === 2
+      ? minuteSuffix.value?.trim() || null
+      : null;
+  }
+  if (part === "secondSuffix") {
+    const secondIndex = parts.indexOf(parts.find(({ type }): boolean => type === "second"));
+    const secondSuffix = parts[secondIndex + 1];
+    return secondSuffix && secondSuffix.type === "literal" ? secondSuffix.value?.trim() || null : null;
+  }
+  return parts.find(({ type }) => (part == "meridiem" ? type === "dayPeriod" : type === part))?.value || null;
+}
+
+export function getMeridiem(hour: string): Meridiem {
+  if (!isValidNumber(hour)) {
+    return null;
+  }
+  const hourAsNumber = parseInt(hour);
+  return hourAsNumber >= 0 && hourAsNumber <= 11 ? "AM" : "PM";
+}
+
+export function isValidTime(value: string): boolean {
+  if (!value || value.startsWith(":") || value.endsWith(":")) {
+    return false;
+  }
   const splitValue = value.split(":");
-  if (splitValue.length > 1) {
+  if (splitValue.length > 1 && splitValue.length < 4) {
```

Comment:
Suggestion: For readability, consider inverting the condition so that the shorter block is first. In this case, this also allows us to reduce nesting. You can also return the value of the last conditional instead of having it fall through:

```ts
const validLength = splitValue.length > 1 && splitValue.length < 4;
if (!validLength) {
  return false;
}

const [hour, minute, second] = splitValue;
// ...
return (hourValid && minuteValid && !second) || (hourValid && minuteValid && secondValid);
```